### PR TITLE
KAFKA-4443; Controller should send UpdateMetadataRequest prior to LeaderAndIsrRequest during failover

### PR DESCRIPTION
**Pull request by lindong28**

### CODE REVIEW

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 1]
```diff
@@ -335,7 +336,6 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
       maybeTriggerPartitionReassignment()
       maybeTriggerPreferredReplicaElection()
       /* send partition leadership info to all live brokers */
-      sendUpdateMetadataRequest(controllerContext.liveOrShuttingDownBrokerIds.toSeq)
```
Comment (becketqin) [inline core/src/main/scala/kafka/controller/KafkaController.scala:338]: Should we still send `UpdateMetadataRequest`s here? The requests sent in line 330 were just include the information stored in ZK before the leader elections. So it seems we still need to propagate the state change (if there is any) in `replicaStateMachine.startup()` and `partitionStateMachine.startup()`.

Comment (lindong28) [reply core/src/main/scala/kafka/controller/KafkaController.scala:338]: @becketqin I think we don't need this. When controller start replica statement and partition and partition state machine, it should send LeanderAndIsrRequest for every partition that will have a leader. `addLeaderAndIsrRequestForBrokers()` will in turn call `addUpdateMetadataRequestForBrokers` to send MetadataUpdateRequest to all live brokers for these partitions. For partition that doesn't have a leader due to NoReplicaOnlineException, MetadataUpdateRequest is not needed for this partition either.

Comment (becketqin) [reply core/src/main/scala/kafka/controller/KafkaController.scala:338]: Yes, you are right. Then it is fine.

Comment (becketqin) [reply core/src/main/scala/kafka/controller/KafkaController.scala:338]: It would probably be useful to add a comment explaining that the order matters.

Comment (lindong28) [reply core/src/main/scala/kafka/controller/KafkaController.scala:338]: Sure. I will add the comment.

**Comment by becketqin** [comment]
LGTM. 

@junrao @jjkoshy Will you also be able to take a look? I took a quick look at KAFKA-901 which introduced sending UpdateMetadataRequest from controller but did not find specific reason on the order. Is there any concern on this change? Thanks.

**Comment by junrao** [comment]
@lindong28 : Thanks for the patch. LGTM. Left a comment in the jira. Perhaps you could clarify there. In KAFKA-3042, there is also discussion on combining UpdateMetadataRequest and LeaderAndIsrRequest together. I am wondering if you have thought about whether that's a better long term solution.

**Comment by ijuma** [comment]
Is it possible to write a test for this (yes, it may sometimes sound like I ask the same question over and over, but it's an important question :)).

**Comment by becketqin** [comment]
@ijuma It might be a little tricky to write test in this case. The bug was caught by a race condition that the broker comes up during the controller fail over and before the controller finishing initializing its context. We can write a test but the test might also subject to race condition unless we put special logic in the actual code to guarantee the order, which seems intrusive. Do you have any suggestion on how to test such cases?

**Comment by ijuma** [comment]
@becketqin I think for cases like this one would probably write test cases using mocks that validate that things happen in a certain order. I understand that the controller doesn't make this easy and it shouldn't be a blocker if it turns out that it's too hard with the current design. But the default should be to include tests with bug fixes and if we don't, there should be an explanation on the PR.

**Pull request closed without merging**